File: backend/app/schemas/voice_live.py
"""Voice Live API request/response schemas."""

from pydantic import BaseModel


class VoiceLiveTokenResponse(BaseModel):
    """Voice Live configuration metadata. Token is always masked (auth handled server-side)."""

    endpoint: str
    token: str
    region: str
    model: str
    avatar_enabled: bool
    avatar_character: str
    voice_name: str
    auth_type: str = "key"  # "key" for API key, "bearer" for STS bearer token
    agent_id: str | None = None
    agent_version: str | None = None
    project_name: str | None = None
    agent_mode_available: bool = False
    agent_warning: str | None = None

    # Per-HCP fields (D-08)
    avatar_style: str = "casual"
    avatar_customized: bool = False
    voice_type: str = "azure-standard"
    voice_temperature: float = 0.9
    voice_custom: bool = False
    turn_detection_type: str = "server_vad"
    noise_suppression: bool = False
    echo_cancellation: bool = False
    eou_detection: bool = False
    recognition_language: str = "auto"


class VoiceLiveModelInfo(BaseModel):
    """Single Voice Live model entry."""

    id: str
    label: str
    tier: str
    description: str


class VoiceLiveModelsResponse(BaseModel):
    """List of supported Voice Live models grouped by tier."""

    models: list[VoiceLiveModelInfo]


class VoiceLiveConfigStatus(BaseModel):
    """Voice Live and Avatar availability status for the current deployment."""

    voice_live_available: bool
    avatar_available: bool
    voice_name: str
    avatar_character: str


class AvatarCharacterStyle(BaseModel):
    """A single style variant of an avatar character."""

    id: str
    display_name: str


class AvatarCharacterInfo(BaseModel):
    """Metadata for one Azure TTS Avatar character."""

    id: str
    display_name: str
    gender: str
    is_photo_avatar: bool = False
    styles: list[AvatarCharacterStyle]
    default_style: str
    thumbnail_url: str


class AvatarCharactersResponse(BaseModel):
    """List of available avatar characters with metadata."""

    characters: list[AvatarCharacterInfo]

File: frontend/src/types/voice-live.ts
export type SessionMode =
  | "text"
  | "voice_pipeline"
  | "digital_human_pipeline"
  | "voice_realtime_model"
  | "digital_human_realtime_model"
  | "voice_realtime_agent"
  | "digital_human_realtime_agent";

export interface VoiceLiveToken {
  endpoint: string;
  token: string;
  auth_type?: "key" | "bearer"; // "key" for API key, "bearer" for STS bearer token
  region: string;
  model: string;
  avatar_enabled: boolean;
  avatar_character: string;
  voice_name: string;
  agent_id?: string;
  agent_version?: string;
  project_name?: string;
  // Per-HCP fields from token broker (D-08)
  avatar_style?: string;
  avatar_customized?: boolean;
  voice_type?: string;
  voice_temperature?: number;
  voice_custom?: boolean;
  turn_detection_type?: string;
  noise_suppression?: boolean;
  echo_cancellation?: boolean;
  eou_detection?: boolean;
  recognition_language?: string;
  agent_instructions_override?: string;
}

export interface VoiceLiveConfigStatus {
  voice_live_available: boolean;
  avatar_available: boolean;
  voice_name: string;
  avatar_character: string;
}

export interface VoiceLiveModelInfo {
  id: string;
  label: string;
  tier: string;
  description: string;
}

export interface VoiceLiveModelsResponse {
  models: VoiceLiveModelInfo[];
}

export type VoiceConnectionState =
  | "disconnected"
  | "connecting"
  | "connected"
  | "reconnecting"
  | "error";

export type AudioState = "idle" | "listening" | "speaking" | "muted";

export interface TranscriptSegment {
  id: string;
  role: "user" | "assistant";
  content: string;
  isFinal: boolean;
  timestamp: number;
}

export interface VoiceLiveOptions {
  language: string;
  systemPrompt: string;
  onTranscript?: (segment: TranscriptSegment) => void;
  onConnectionStateChange?: (state: VoiceConnectionState) => void;
  onAudioStateChange?: (state: AudioState) => void;
  onError?: (error: Error) => void;
}

export interface VoiceLiveControls {
  connect: (
    hcpProfileId: string,
    systemPrompt?: string,
  ) => Promise<{
    avatarEnabled: boolean;
    model: string;
    iceServers: RTCIceServer[];
  }>;
  disconnect: () => Promise<void>;
  toggleMute: () => void;
  sendTextMessage: (text: string) => Promise<void>;
  /** Send audio data (PCM16 base64-encoded) via backend proxy. */
  sendAudio: (base64Audio: string) => void;
  /** Send raw Azure RT protocol message via backend proxy. */
  send: (data: unknown) => void;
  isMuted: boolean;
  connectionState: VoiceConnectionState;
  audioState: AudioState;
  avatarSdpCallbackRef: React.MutableRefObject<
    ((serverSdp: string) => void) | null
  >;
}

/** Runtime voice configuration settings for the config panel. */
export interface VoiceConfigSettings {
  /** Speech input language code (e.g. "zh-CN", "en-US") or "auto" for auto-detect. */
  language: string;
  /** Whether to auto-detect the spoken language. */
  autoDetect: boolean;
  /** Show partial AI responses while still generating. */
  interimResponse: boolean;
  /** AI initiates conversation proactively. */
  proactiveEngagement: boolean;
}

export interface AvatarStreamControls {
  /** Start avatar WebRTC handshake. Sends SDP offer via VoiceLive session event. */
  connect: (
    iceServers: RTCIceServer[],
    sendSdpOffer: (sdp: string) => Promise<void>,
  ) => Promise<void>;
  /** Handle SDP answer from server (via onSessionAvatarConnecting handler). */
  handleServerSdp: (serverSdp: string) => Promise<void>;
  disconnect: () => void;
  isConnected: boolean;
}
